src/header/header_def.rs

```diff
@@ -0,0 +1,446 @@
+
+                    while current_offset < end {
+                        // Padding
+                        if raw_packet[current_offset] == 0x00 {
+                            current_offset += 1;
+                            continue;
+                        }
+
+                        let ext_id = raw_packet[current_offset] >> 4;
+                        let len = (raw_packet[current_offset] as usize & !0xF0) + 1;
+                        current_offset += 1;
+
+                        if ext_id == super::EXTENSION_ID_RESERVED {
+                            break;
+                        }
+
+                        self.extensions.push(super::Extension {
+                            id: ext_id,
+                            payload: raw_packet[current_offset..current_offset + len].to_vec(),
+                        });
+
+                        current_offset += len;
+                    }
+                }
+
+                // RFC 8285 RTP Two Byte Header Extension
+                super::ExtensionProfile::TwoByte => {
+                    let end = current_offset + extension_length as usize;
+
+                    while current_offset < end {
+                        // Padding
+                        if raw_packet[current_offset] == 0x00 {
+                            current_offset += 1;
+                            continue;
+                        }
+
+                        let ext_id = raw_packet[current_offset];
+                        current_offset += 1;
+
+                        let len = raw_packet[current_offset];
+                        current_offset += 1;
+
+                        self.extensions.push(super::Extension {
+                            id: ext_id,
+                            payload: raw_packet[current_offset..current_offset + len as usize]
+                                .to_vec(),
+                        });
+
+                        current_offset += len as usize;
+                    }
+                }
+
+                // RFC3550 Extension
+                _ => {
+                    if raw_packet.len() < current_offset + extension_length as usize {
+                        return Err(RTPError::HeaderSizeInsufficientForExtension);
+                    }
+
+                    self.extensions.push(super::Extension {
+                        id: 0,
+                        payload: raw_packet
+                            [current_offset..current_offset + extension_length as usize]
+                            .to_vec(),
+                    });
+
+                    current_offset += self.extensions[0].payload.len();
+                }
+            }
+        }
+
+        Ok(current_offset)
+    }
+
+    /// Marshal serializes the packet into bytes.
+    pub fn marshal(&mut self) -> Result<Vec<u8>, RTPError> {
+        let mut buf = vec![0u8; self.marshal_size()];
+
+        let size = self.marshal_to(&mut buf)?;
+        buf.truncate(size);
```

Comment:
do we have such chance to truncate buf size due to inaccurate self.marshal_size() calculation?

Reply:
Yes, it's possible one of the header field is mutated adding extra bits to marshal_size. We should make extra checks to ensure there's no padded bits